pkg/controllers/updaterun/execution.go

```diff
@@ -0,0 +1,482 @@
+	toBeDeletedBindings []*placementv1beta1.ClusterResourceBinding,
+) (bool, error) {
+	updateRunRef := klog.KObj(updateRun)
+	existingDeleteStageStatus := updateRun.Status.DeletionStageStatus
+	existingDeleteStageClusterMap := make(map[string]*placementv1alpha1.ClusterUpdatingStatus, len(existingDeleteStageStatus.Clusters))
+	for i := range existingDeleteStageStatus.Clusters {
+		existingDeleteStageClusterMap[existingDeleteStageStatus.Clusters[i].ClusterName] = &existingDeleteStageStatus.Clusters[i]
+	}
+	deletingBinding := 0
+	for _, binding := range toBeDeletedBindings {
+		curCluster, exist := existingDeleteStageClusterMap[binding.Spec.TargetCluster]
+		if !exist {
+			// This is unexpected because we already checked in validation.
+			missingErr := controller.NewUnexpectedBehaviorError(fmt.Errorf("the to be deleted cluster `%s` is not in the deleting stage during execution", binding.Spec.TargetCluster))
+			klog.ErrorS(missingErr, "The cluster in the deleting stage does not include all the to be deleted binding", "clusterStagedUpdateRun", updateRunRef)
+			return false, fmt.Errorf("%w: %s", errStagedUpdatedAborted, missingErr.Error())
+		}
+		// In validation, we already check the binding must exist in the status.
+		delete(existingDeleteStageClusterMap, binding.Spec.TargetCluster)
+		// Make sure the cluster is not marked as deleted as the binding is still there.
+		if condition.IsConditionStatusTrue(meta.FindStatusCondition(curCluster.Conditions, string(placementv1alpha1.ClusterUpdatingConditionSucceeded)), updateRun.Generation) {
+			unexpectedErr := controller.NewUnexpectedBehaviorError(fmt.Errorf("the deleted cluster `%s` in the deleting stage still has a clusterResourceBinding", binding.Spec.TargetCluster))
+			klog.ErrorS(unexpectedErr, "The cluster in the deleting stage is not removed yet but marked as deleted", "cluster", curCluster.ClusterName, "clusterStagedUpdateRun", updateRunRef)
+			return false, fmt.Errorf("%w: %s", errStagedUpdatedAborted, unexpectedErr.Error())
+		}
+		if condition.IsConditionStatusTrue(meta.FindStatusCondition(curCluster.Conditions, string(placementv1alpha1.ClusterUpdatingConditionStarted)), updateRun.Generation) {
+			// The cluster status is marked as being deleted.
+			if binding.DeletionTimestamp.IsZero() {
+				// The cluster is marked as deleting but the binding is not deleting.
+				unexpectedErr := controller.NewUnexpectedBehaviorError(fmt.Errorf("the cluster `%s` in the deleting stage is marked as deleting but its corresponding binding is not deleting", curCluster.ClusterName))
+				klog.ErrorS(unexpectedErr, "The binding should be deleting before we mark a cluster deleting", "clusterStatus", curCluster, "clusterStagedUpdateRun", updateRunRef)
+				return false, fmt.Errorf("%w: %s", errStagedUpdatedAborted, unexpectedErr.Error())
+			}
+			deletingBinding++
+			continue
+		}
+		// The cluster status is not deleting yet
+		if err := r.Client.Delete(ctx, binding); err != nil {
```

Comment:
Compared with the original implementation, I did not check if binding is up-to-date before deleting. I don't think it's necessary. Do we want that?

Reply:
In case of both resource changes and scheduling policy changes, the binding to be deleted may not have up-to-date resources. If there are concurrent update runs, the same binding could be deleted by either updateRun. If there's a new scheduling policy change that marks this to-be-deleted binding as scheduled again, this updateRun can fail during validation. Or if there's chance that the change happens right after the validation passes and before deleting the bindings, I think I can add a check to make sure the binding's state is unScheduled or report unexpected error if not.

Reply:
I do have a concern: in rollout controller, we make sure to complete the deletion of the bindings on a cluster before scheduling a newly created binding on the same cluster: https://github.com/Azure/fleet/blob/main/pkg/controllers/rollout/controller.go#L226. I wonder if we should have similar logic in the updateRun controller too.

Reply:
the concurrent updateRun is not valid if the schedule results are different. This is only for the case one needs to release different versions of the resources so the bindings should all be scheduled or bound.

Reply:
As for the check no duplicate logic, should the check at https://github.com/Azure/fleet/blob/cb0c5f70b850322a221a70c18d51008ffe43f447/pkg/controllers/updaterun/initialization.go#L177 be suffice?

Reply:
_"the concurrent updateRun is not valid if the schedule results are different. This is only for the case one needs to release different versions of the resources so the bindings should all be scheduled or bound."_

In this case, the bindings won't be marked as to-be-deleted because they both have the latest scheduling snapshot index.